### ISSUE TITLE
Configuration of built-in snippets

### ISSUE THREAD

**Issue by echasnovski** [neovim/neovim#30306]
### Problem

Currently there is no way to configure built-in `vim.snippet` behavior. This recently became a problem after introduction of not configurable "session only" mappings for snippet navigation (see #30198).

### Expected behavior

Some way for users to configure how certain aspects of snippets expansion is done. As per [this comment](https://github.com/neovim/neovim/issues/30198#issuecomment-2336620609), it might be a good idea to define the scope before deciding on this.

Here are some suggestions:
- Implement `vim.snippet.config()`, similar in spirit to `vim.diagnostic.config()`. Currently this seems like the best (future-proof, already familiar) design. See also [this comment](https://github.com/neovim/neovim/issues/30198#issuecomment-2336624245) for some counter-points.
- Allow `opts` in `vim.snippet.expand()`. This allows configuration on a call level which might be good for plugins that leverage it. However, it also might be too granular to the point of not solving the issue for built-in functionality (see [this comment](https://github.com/neovim/neovim/issues/30198#issuecomment-2336583394)).
- Implement `SnippetEnter` and `SnippetLeave` events (see #26449) for users to be able to perform actions only inside snippet session. Not *exactly* the snippet configuration, but at least *some* way for users to act on snippet expansion.

**Comment by MariaSolOs** [comment]
Since the current snippet settings aren't specific to a session but instead configure the entire snippet expansion logic, using `opts` in `vim.snippet.expand` might have a lot of "global" properties that IMO don't really belong to that function. I would lean towards `vim.snippet.config`, with maybe a `on_expand` callback to include the behavior from the former point.

**Comment by mfussenegger** [comment]
> Currently this seems like the best (future-proof, already familiar) design

I find this assessment a bit odd given the points already made in the linked issue:

- `vim.diagnostic.config` is the exception, not the rule in core. There are global/local buffer/window options for most of the other configurations. `diagnostic.config` is the exception because it required namespace scoping. That's not required for snippets.
- It would require a keymap DSL, which was already identified as sub-optimal in other issues/plugin guide PRs.

> Allow opts in vim.snippet.expand(). This allows configuration on a call level which might be good for plugins that leverage it. However, it also might be too granular to the point of not solving the issue for built-in functionality (see this comment).

Maybe I haven't been clear enough in the other issue. This is a no-go for me. I don't see us adding a `vim.completion.enable({ snippet_keymap =  })` _and_ a `vim.lsp.buf.code_action({ snippet_keymap = ... })` _and_ `vim.lsp.future_snippet_use_case...`

Besides spreading the configuration to each function that triggers `vim.snippet.expand` under the hood, it also suffers from the keymap-DSL problem, unless we'd make it `on_snippet_start` triggers, but at that point we'd be emulating snippet autocmds.

---

Please let us not introduce new ways of doing things. 
This should either be:

- A global keybinding instead of context dependent that's set as default which users can override like other default keybindings.
- https://github.com/neovim/neovim/issues/26449

**Comment by echasnovski** [comment]
@mfussenegger, this issue is **not** about not configurable keymaps. They are mentioned as a driver for creating this and discussed in #30198.

This is meant as a place for a general discussion (as per [this comment](https://github.com/neovim/neovim/issues/30198#issuecomment-2336620609)). Hence locating all suggested approaches in one place for reference (however strongly you put some of them as a no-go for you).

**Comment by clason** [comment]
Yes, but the discussion then becomes (too) quickly about keymaps. The point of such an issue would be to collect _what_ to configure before starting to propose _how_.
